### ISSUE TITLE
Race condition during elastic agent assignment 

### ISSUE THREAD

**Issue by jyotisingh** [gocd/gocd#3095]
##### Issue Type

- Bug Report

##### Summary
 This was seen with a couple of docker swarm elastic agent, but would apply to all elastic agents.

Scenario:
* Say an elastic agent (EA1) was brought up to build a job (say Job1), the job execution finished and now the agent has been marked as idle. 
* Meanwhile another job(say Job2) got scheduled which has exactly the same agent requirements (ie. elastic-plugin-profile being the same). 
* A periodic ping request is sent by the server to the plugin. Plugin requests for the status of its agents, EA1 is reported as idle. Plugins initiates actions to bring down the agent - ie. disable, delete from config, and terminate instance.
* Before the disable action finishes off, if the Server gets a ping from EA1, it would find that Job2 can be assigned to it, so it asks the plugin if it should go ahead with the assignment. Plugin checks that the agent properties matches the job and answers yes (it should probably have checked if it has already started to cleanup the agent but it doesn't do that right now), so the assignment is done. 
* After the assignment, the disable, delete and agent termination happens. The job gets abandoned and is then rescheduled. 

##### Environment

This is happening often on build.gocd.io. The console log itself wouldn't show this as rerun as the agent went down before the job execution could start on the agent. Check the db records for instances:
https://build.gocd.io/go/tab/build/detail/build-linux-PR/1121/build-server/1/jetty9
https://build.gocd.io/go/tab/build/detail/regression-PR/153/regression-linux/1/fanin-tests-runInstance-1

**Comment by ketan** [comment]
@jyotisingh is this fixed?

**Comment by jyotisingh** [comment]
No @ketan. I had picked it up three weeks ago but fell sick the very next day. Then I was traveling and then completely forgot about it. Oops. Thanks for reminding. But I had picked it up just coz I had nothing else to work on. Did you need this anything?

**Comment by ketan** [comment]
> Did you need this anything?

No, I thought that maybe we fixed it, but forgot to close it out.

**Comment by GaneshSPatil** [comment]
Closed it by mistake. Reopening it.
